lualib/rspamadm/mime.lua

```diff
@@ -873,19 +879,85 @@ local function sign_handler(opts)
   end
 end
 
+-- Strips directories and .extensions (if present) from a filepath
+  -- very_simple
+  -- /home/very_simple.eml
+  -- very_simple.eml
+  -- very_simple.example.eml
+  -- /home/very_simple
+  -- home/very_simple
+  -- ./home/very_simple
+  -- ../home/very_simple.eml
+--All the above end up as very_simple
+local function filename_only(filepath)
+  filename = filepath:match(".*%/([^%.]+)")
+  if not filename then
+    filename = filepath:match("([^%.]+)")
+  end
+  return filename
+end
+
+--Write the dump content to file or standard out
+local function write_dump_content(dump_content, fname, extension, outdir)
+  
+  wrote_filepath = nil
+  if outdir then
+    if outdir:sub(-1) ~= "/" then
+      outdir = outdir .. "/"
+    end
+
+    local outpath = string.format("%s%s.%s", outdir, filename_only(fname), extension)
+    local outfile = io.open(outpath, "w")
+
+    if outfile then
+      outfile:write(dump_content)
+      outfile:close()
+      io.write(outpath.."\n")
+      wrote_filepath = outpath
+    else
+      io.stderr:write(string.format("Unable to open: %s\n", outpath))
+    end
+  else
+      io.write(dump_content)
+  end
+
+  return wrote_filepath
+end
+
+-- Get the formatted ucl (split or unsplit) or the raw task content
+local function get_dump_content(task, opts, fname)
+  if opts.ucl or opts.json or opts.messagepack then
+    local ucl_object = lua_mime.message_to_ucl(task)
+    
+    --Split out the content field into separate raws and update the ucl
+    if opts.split then
+      for i, part in ipairs(ucl_object.parts) do
+        if part.content then
+          local part_filename = string.format("%s-part%d", filename_only(fname), i)
+          local part_path = write_dump_content(tostring(part.content), part_filename, "raw", opts.outdir)
```

Comment:
I don't understand why do we call `tostring` here. It will allocate and intern a potentially large string in Lua. It would be better to adopt `write_dump_content` function for the `rspamd_text` case.

Reply:
Further down the stack I was getting errors from file write because the type wasn't a string it was userdata. A quick google explained that tostring could be used if the type implemented it (via something like __string - i forget). I tried it, it worked so I stopped there.

If there's a better way please point me at some examples. I am not very good with lua.

Reply:
Yes, so I would suggest to add a simple `if` to check that `type(data) == 'string'` or `type(data) == 'userdata'`. For `rspamd_text` you could use [save_in_file](https://rspamd.com/doc/lua/rspamd_text.html#m0d82f) method. I don't understand also, why do you add `\n` at the end.

Reply:
added if type check before tostring calls